core/lib/src/fs/server.rs

```diff
@@ -60,351 +72,492 @@ use crate::fs::NamedFile;
+impl<F> Rewriter for MapFile<F>
+    where F: for<'p, 'h> Fn(File<'p, 'h>) -> FileResponse<'p, 'h> + Send + Sync + 'static,
+{
+    fn rewrite<'p, 'h>(&self, path: Option<FileResponse<'p, 'h>>) -> Option<FileResponse<'p, 'h>> {
         match path {
-            Some(p) if p.is_dir() => {
-                // Normalize '/a/b/foo' to '/a/b/foo/'.
-                if options.contains(Options::NormalizeDirs) && !req.uri().path().ends_with('/') {
-                    let normal = req.uri().map_path(|p| format!("{}/", p))
-                        .expect("adding a trailing slash to a known good path => valid path")
-                        .into_owned();
-
-                    return Redirect::permanent(normal)
-                        .respond_to(req)
-                        .or_forward((data, Status::InternalServerError));
-                }
-
-                if !options.contains(Options::Index) {
-                    return Outcome::forward(data, Status::NotFound);
-                }
-
-                let index = NamedFile::open(p.join("index.html")).await;
-                index.respond_to(req).or_forward((data, Status::NotFound))
-            },
-            Some(p) => {
-                let file = NamedFile::open(p).await;
-                file.respond_to(req).or_forward((data, Status::NotFound))
-            }
-            None => Outcome::forward(data, Status::NotFound),
+            Some(FileResponse::File(file)) => Some(self.0(file)),
+            path => path,
         }
     }
 }
 
-/// A bitset representing configurable options for [`FileServer`].
+/// Prepends the provided path, to serve files from a directory.
+///
+/// You can use [`relative!`] to make a path relative to the crate root, rather
+/// than the runtime directory.
 ///
-/// The valid options are:
+/// # Example
+///
+/// ```rust,no_run
+/// # use rocket::fs::{FileServer, dir_root, relative};
+/// # fn make_server() -> FileServer {
+/// FileServer::empty()
+///     .map_file(dir_root(relative!("static")))
+/// # }
+/// ```
 ///
-///   * [`Options::None`] - Return only present, visible files.
-///   * [`Options::DotFiles`] - In addition to visible files, return dotfiles.
-///   * [`Options::Index`] - Render `index.html` pages for directory requests.
-///   * [`Options::IndexFile`] - Allow serving a single file as the index.
-///   * [`Options::Missing`] - Don't fail if the path to serve is missing.
-///   * [`Options::NormalizeDirs`] - Redirect directories without a trailing
-///     slash to ones with a trailing slash.
+/// # Panics
 ///
-/// `Options` structures can be `or`d together to select two or more options.
-/// For instance, to request that both dot files and index pages be returned,
-/// use `Options::DotFiles | Options::Index`.
-#[derive(Debug, Clone, Copy)]
-pub struct Options(u8);
+/// Panics if `path` is not directory.
+pub fn dir_root(path: impl AsRef<Path>)
+    -> impl for<'p, 'h> Fn(File<'p, 'h>) -> FileResponse<'p, 'h> + Send + Sync + 'static
```

Comment:
This will be hard to understand. We _should_ be able to return `impl Rewriter`, I believe, which would result in a much simpler type signature.

Reply:
While we could, the return type would not be useable in map_file or filter_file. The only way to implement a return type that support both and_rewrite(impl Rewriter) and map_file(impl Fn(file) -> Response) would be to add traits for MapFile and FilterFile, so a custom struct could implement both. I thought this would complicate the API more, and choose to leave it with a more complex return type for now. I'm not sure what the right way to go here is, but I'm leaning towards adding MapFile and FilterFile traits.

Reply:
After doing some testing, I've updated this to return `impl FileMap`, which is more or less a type alias of the full `impl Fn...`. I initially wanted to make them proper traits, but it would prevent using lambdas in a convenient way, since the Rust type inferencing can't automatically determine the HRTBs.